src/attribute/ruby_glue.rs

```diff
@@ -569,27 +546,57 @@ extern "C" fn init_with_from_user(this: ffi::VALUE, coder: ffi::VALUE) -> ffi::V
         let this = get_struct_mut::<Attribute>(this);
         init_with_populated(this, coder);
         let original_attribute = ffi::rb_funcall(coder, id!("[]"), 1, rstr!("original_attribute"));
-        let original_attribute = if ffi::RB_NIL_P(original_attribute) {
-            None
+        let new_source = if ffi::RB_NIL_P(original_attribute) {
+            // Even though this was a `FromUser` subclass, if this YAML was
+            // dumped in Rails 4.2, `original_attribute` won't be there.
+            // `UserProvidedDefault` is the only thing that can have no original.
+            Source::UserProvidedDefault(None)
         } else {
-            Some(Box::new(
-                get_struct::<Attribute>(original_attribute).clone(),
+            Source::FromUser(Box::new(
+                get_struct::<Attribute>(original_attribute).clone()
             ))
         };
-        // Even though this was a `FromUser` subclass, if this YAML was
-        // dumped in Rails 4.2, `original_attribute` won't be there.
-        // `UserProvidedDefault` is the only thing that can have no original.
+
         match *this {
             Attribute::Populated { ref mut source, .. } => {
-                *source = Source::UserProvidedDefault(original_attribute)
-            }
+                *source = new_source
+            },
             _ => unreachable!(),
         }
 
         ffi::Qnil
     }
 }
 
+extern "C" fn init_with(this: ffi::VALUE, coder: ffi::VALUE) -> ffi::VALUE {
+    unsafe {
+        use std::slice;
+        use std::ffi::CString;
+        use std::str;
+
+        let tag = ffi::rb_funcall(coder, id!("tag"), 0);
+        let tag_ptr = ffi::RSTRING_PTR(tag);
+        let tag_len = ffi::RSTRING_LEN(tag);
+        let tag = slice::from_raw_parts(tag_ptr as _, tag_len as usize);
+
+        let tag_prefix_len = b"!ruby/object:ActiveModel::Attribute::".len();
+
+        match &tag[tag_prefix_len..] {
+            b"FromUser" | b"UserProvidedDefault" => init_with_from_user(this, coder),
+            b"FromDatabase" => init_with_from_database(this, coder),
+            b"WithCastValue" => init_with_precast(this, coder),
+            b"Uninitialized" => init_with_uninitialized(this, coder),
+            other => {
+                let tag_str = str::from_utf8(other).unwrap_or("{invalid UTF-8}");
+                let error = format!("Unrecognized tag: {}", tag_str);
+                // This trunctates the tag if there is a 0 byete in the middle of our Rust string.
+                let message = CString::from_vec_unchecked(error.into());
```

Comment:
I'm unclear why we are going through `CString` for this

Reply:
Also you should never ever ever ever ever use methods with the word `unchecked` in the name (or unsafe functions in general) unless you thoroughly understand the preconditions it's expecting and have documented why they're upheld at the call site.

In this case the function is doing the exact opposite of what you describe, this very specifically doesn't do what you described, it does the exact opposite.